### ISSUE TITLE
Conntrack entry mismatch leads to policy enforcement on reply packet for service loopback case

### ISSUE THREAD

**Issue by skmatti** [cilium/cilium#19406]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Steps to reproduce:

```
minikube start --network-plugin=cni --cni=false --memory=4096 --kubernetes-version=v1.22.8

cilium install --version=v1.10.5

kubectl apply -f - <<EOF
apiVersion: v1
kind: ReplicationController
metadata:
  name: guestbook
  labels:
    k8s-app.guestbook: web
spec:
  replicas: 1
  selector:
    k8s-app.guestbook: web
  template:
    metadata:
      labels:
        k8s-app.guestbook: web
    spec:
      containers:
      - image: gcr.io/google-samples/gb-frontend:v6
        name: guestbook
        ports:
        - containerPort: 80
          name: http-server
          protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: guestbook
  labels:
    k8s-app.guestbook: web
spec:
  ports:
  - port: 81
    protocol: TCP
    targetPort: http-server
  selector:
    k8s-app.guestbook: web
  type: ClusterIP
EOF

kubectl apply -f - <<EOF
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: allow-ingress-on-tcp-80
specs:
- endpointSelector:
    matchLabels:
      k8s-app.guestbook: web
  ingress:
  - toPorts:
    - ports:
      - port: "80"
        protocol: TCP
EOF
```

It is important that the service port (81) is different from the container port (80). Issue does not occur if they are the same.

```
$ kubectl get svc guestbook
NAME        TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
guestbook   ClusterIP   10.98.138.60   <none>        81/TCP    42m
$ kubectl get pods -o wide --selector=k8s-app.guestbook=web
NAME              READY   STATUS    RESTARTS   AGE   IP           NODE       NOMINATED NODE   READINESS GATES
guestbook-8b46v   1/1     Running   0          43m   10.0.0.208   minikube   <none>           <none>
```

Send traffic to itself via cluster IP:
```
$ kubectl exec guestbook-8b46v -- curl 10.98.138.60:81
```

Cilium monitor output:
```
$ cilium monitor --related-to 1221
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
-> endpoint 1221 flow 0x95218252 identity 22887->22887 state established ifindex lxccbaae0c0d570 orig-ip 169.254.42.1: 169.254.42.1:55518 -> 10.0.0.208:80 tcp SYN
Policy verdict log: flow 0xfef8e31c local EP ID 1221, remote ID 22887, proto 6, ingress, action deny, match none, 10.98.138.60:81 -> 10.0.0.208:55518 tcp SYN, ACK
xx drop (Policy denied) flow 0xfef8e31c to endpoint 1221, identity 22887->22887: 10.98.138.60:81 -> 10.0.0.208:55518 tcp SYN, ACK
Policy verdict log: flow 0x8caf00f9 local EP ID 1221, remote ID 22887, proto 6, ingress, action deny, match none, 10.98.138.60:81 -> 10.0.0.208:55518 tcp SYN, ACK
```

As per my understanding, a pod sending packets to itself via cluster IP should have skipped policy enforcement all together. Although, the first packet (pod -> itself via cluster IP) is skipped policy enforcement, the reply packet is not skipped (incorrectly subjected to policy enforcement). This is happening because the reply packet do not match any conntrack entries.

```
$ cilium bpf ct list global | grep 10.0.0.208
TCP IN 169.254.42.1:55522 -> 10.0.0.208:80 expires=17529096 RxPackets=3 RxBytes=222 RxFlagsSeen=0x02 LastRxReport=17529035 TxPackets=6 TxBytes=444 TxFlagsSeen=0x12 LastTxReport=17529035 Flags=0x0008 [ LBLoopback ] RevNAT=7 SourceSecurityID=22887 IfIndex=0
TCP OUT 10.0.0.208:55522 -> 10.98.138.60:80 expires=17529096 RxPackets=0 RxBytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=3 TxBytes=222 TxFlagsSeen=0x02 LastTxReport=17529035 Flags=0x0008 [ LBLoopback ] RevNAT=7 SourceSecurityID=22887 IfIndex=0
TCP OUT 10.98.138.60:81 -> 10.0.0.208:55522 service expires=17529096 RxPackets=0 RxBytes=7 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=0 TxBytes=0 TxFlagsSeen=0x02 LastTxReport=17529035 Flags=0x0000 [ ] RevNAT=7 SourceSecurityID=0 IfIndex=0
ICMP OUT 10.0.0.208:0 -> 10.98.138.60:0 related expires=17529093 RxPackets=0 RxBytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=1 TxBytes=74 TxFlagsSeen=0x02 LastTxReport=17529035 Flags=0x0018 [ LBLoopback SeenNonSyn ] RevNAT=7 SourceSecurityID=22887 IfIndex=0
```

As per my understanding, the second entry is causing the problem:
`10.0.0.208:55522 -> 10.98.138.60:80` - the destination IP address is the Cluster IP but the port is container port (80) instead of service port (81).

The reply packet on the other hand has `10.98.138.60:81 -> 10.0.0.208:55518` which does not match the conntrack entry and is incorrectly subjected to network policy enforcement.

### Cilium Version

1.10.5

### Kernel Version

x86_64 

### Kubernetes Version

1.22.8

### Sysdump

[cilium-sysdump-20220411-230655.zip](https://github.com/cilium/cilium/files/8470117/cilium-sysdump-20220411-230655.zip)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by skmatti** [comment]
This issue related to https://github.com/cilium/cilium/issues/11593 and https://github.com/cilium/cilium/pull/15321

**Comment by pchaigno** [comment]
Did you check if the same bug also affects master?

**Comment by skmatti** [comment]
Yes, I tried `v1.12.0-rc1` which was created Yesterday and it has the issue

**Comment by skmatti** [comment]
I enabled debug logs which confirms that the reply packet is treated as a new packet:

```
CPU 04: MARK 0x41fb41bb FROM 1143 DEBUG: Conntrack lookup 1/2: src=10.105.185.194:81 dst=10.0.0.92:53112
CPU 04: MARK 0x41fb41bb FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=0
CPU 04: MARK 0x41fb41bb FROM 1143 DEBUG: CT verdict: New, revnat=0
CPU 04: MARK 0x41fb41bb FROM 1143 DEBUG: Policy evaluation would deny packet from 2899 to 2899
Policy verdict log: flow 0x41fb41bb local EP ID 1143, remote ID 2899, proto 6, ingress, action deny, match none, 10.105.185.194:81 -> 10.0.0.92:53112 tcp SYN, ACK
```

Cluster IP - 10.105.185.194
Pod IP - 10.0.0.92

**Comment by pchaigno** [comment]
So if I understand correctly, the issue here is that we're not using the correct port when creating the conntrack entry? Did you check the code to see why that is?

**Comment by skmatti** [comment]
I am using v1.12.0-rc1 for debugging. The conntrack entries are created correctly with this version.

Cluster IP - 10.105.185.194
Pod IP - 10.0.0.92
```
$ cilium bpf ct list global
Defaulted container "cilium-agent" out of: cilium-agent, ebpf-mount (init), clean-cilium-state (init)
TCP OUT 10.105.185.194:81 -> 10.0.0.92:53162 service expires=17762859 RxPackets=0 RxBytes=4 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=0 TxBytes=0 TxFlagsSeen=0x02 LastTxReport=17762801 Flags=0x0000 [ ] RevNAT=4 SourceSecurityID=0 IfIndex=0 
TCP OUT 169.254.42.1:53162 -> 10.0.0.92:80 expires=17762859 RxPackets=0 RxBytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=5 TxBytes=370 TxFlagsSeen=0x02 LastTxReport=17762801 Flags=0x0008 [ LBLoopback ] RevNAT=4 SourceSecurityID=2899 IfIndex=0 
TCP IN 169.254.42.1:53162 -> 10.0.0.92:80 expires=17762859 RxPackets=5 RxBytes=370 RxFlagsSeen=0x02 LastRxReport=17762801 TxPackets=10 TxBytes=740 TxFlagsSeen=0x12 LastTxReport=17762801 Flags=0x0008 [ LBLoopback ] RevNAT=4 SourceSecurityID=2899 IfIndex=0 
ICMP OUT 169.254.42.1:0 -> 10.0.0.92:0 related expires=17762843 RxPackets=0 RxBytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxPackets=1 TxBytes=74 TxFlagsSeen=0x02 LastTxReport=17762785 Flags=0x0018 [ LBLoopback SeenNonSyn ] RevNAT=4 SourceSecurityID=2899 IfIndex=0 
```

But the issue is with the conntrack lookup:
```
CPU 04: MARK 0x41fb41bb FROM 1143 DEBUG: Conntrack lookup 1/2: src=10.105.185.194:81 dst=10.0.0.92:53112
CPU 04: MARK 0x41fb41bb FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=0
```

flags=0 which means the lookup was initialized from https://github.com/cilium/cilium/blob/964b034aa91c50a145416ec3a94093e3beb1b06e/bpf/bpf_lxc.c#L664 but it should not have reached this line. Instead, the conntrack lookup should have been initialized from https://github.com/cilium/cilium/blob/964b034aa91c50a145416ec3a94093e3beb1b06e/bpf/bpf_lxc.c#L1078 because it is service packet.

Somehow packet fails lb4_lookup_service test and misses lb4_local.

Does this summary looks accurate?. If so, how do I proceed with determining why service check fails for the reply packet?.

**Comment by pchaigno** [comment]
`flags=0` means `tuple->flags == TUPLE_F_OUT` which itself means `dir == CT_INGRESS`. I think you are looking at https://github.com/cilium/cilium/blob/c5c7d4126bbec0871cff05a2942c4b13d8131cac/bpf/lib/nodeport.h#L1974

**Comment by skmatti** [comment]
I see, thanks. But why would it go through nodeport, it is a cluster IP service. Maybe it is coming from https://github.com/cilium/cilium/blob/964b034aa91c50a145416ec3a94093e3beb1b06e/bpf/bpf_lxc.c#L1522

**Comment by pchaigno** [comment]
Oh, right. That makes more sense.

So it's ingressing into the container and we're doing a CT lookup with the VIP instead of the loopback IP. So why wasn't the packet [SNATed to the loopback IP](https://github.com/cilium/cilium/blob/a0dabcd796b53564a76b4b0a374ce94cb37ce410/bpf/lib/lb.h#L1573-L1589) on egress?

**Comment by skmatti** [comment]
Here is the full conntrack lookups for the trafffic.

There was a lookup with loopback IP but the verdict was `Reply`. Then source NAT and DNAT happens and a conntrack lookup is initiated again:

```
$ kubectl -n kube-system exec cilium-dbvqk -- cilium monitor --related-to 1143 -v
Defaulted container "cilium-agent" out of: cilium-agent, ebpf-mount (init), clean-cilium-state (init)
Listening for events on 12 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack lookup 1/2: src=10.0.0.92:53162 dst=10.105.185.194:81
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=4
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: CT verdict: New, revnat=0
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Service backend slot lookup: slot=2, dport=81
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack create: proxy-port=0 revnat=4 src-identity=0 lb=0.0.0.0
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Loopback SNAT from=10.0.0.92 to=169.254.42.1
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Successfully mapped addr=10.0.0.92 to identity=2899
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack lookup 1/2: src=169.254.42.1:53162 dst=10.0.0.92:80
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: CT verdict: New, revnat=0
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack create: proxy-port=0 revnat=4 src-identity=2899 lb=169.254.42.1
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Attempting local delivery for container id 1143 from seclabel 2899
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack lookup 1/2: src=169.254.42.1:53162 dst=10.0.0.92:80
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=0
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: CT entry found lifetime=17762843, revnat=4
CPU 04: MARK 0xe97cb587 FROM 1143 DEBUG: CT verdict: Established, revnat=4
-> endpoint 1143 flow 0xe97cb587 , identity 2899->2899 state established ifindex lxc9ea2309f4588 orig-ip 169.254.42.1: 169.254.42.1:53162 -> 10.0.0.92:80 tcp SYN
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Successfully mapped addr=169.254.42.1 to identity=2
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Conntrack lookup 1/2: src=10.0.0.92:80 dst=169.254.42.1:53162
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: CT entry found lifetime=17762843, revnat=4
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: CT verdict: Reply, revnat=4
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Reverse NAT lookup, index=4
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Performing reverse NAT, address=10.105.185.194 port=81
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Loopback reverse SNAT from=169.254.42.1 to=10.0.0.92
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Attempting local delivery for container id 1143 from seclabel 2899
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Conntrack lookup 1/2: src=10.105.185.194:81 dst=10.0.0.92:53162
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=0
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: CT verdict: New, revnat=0
CPU 04: MARK 0xee59446 FROM 1143 DEBUG: Policy evaluation would deny packet from 2899 to 2899
Policy verdict log: flow 0xee59446 local EP ID 1143, remote ID 2899, proto 6, ingress, action deny, match none, 10.105.185.194:81 -> 10.0.0.92:53162 tcp SYN, ACK
```

**Comment by joestringer** [comment]
I wonder if the next step could be to create a new test in https://github.com/cilium/cilium-cli connectivity testsuite that configures a service with only one pod backing it, then initiates requests from that pod to the service with policy applied. This could make reproducing the issue more easy, which should also make it easier to debug and figure out a solution.

**Comment by aditighag** [comment]
If I understand the issue correctly, we already have a test case to verify this scenario - https://github.com/cilium/cilium/blob/master/test/k8s/services.go#L170-L170. 

>It is important that the service port (81) is different from the container port (80). Issue does not occur if they are the same.

If this is indeed relevant, we should extend the test case. @skmatti Here is context for how we skip the policy enforcement for the service loopback + policy enforcement case, this hopefully helps in debugging/fixing the issue further.

**Comment by sypakine** [comment]
This may be fixed by https://github.com/cilium/cilium/pull/27798

**Comment by julianwiedmann** [comment]
> This may be fixed by #27798

agree. @skmatti could you please re-test?

**Comment by skmatti** [comment]
The issue is resolved on cilium `1.12.15` so most likely fixed by https://github.com/cilium/cilium/pull/27798

```
-> endpoint 2244 flow 0xabe2de97 , identity 43665->43665 state new ifindex lxc54076bf7fdd8 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp SYN
-> endpoint 2244 flow 0xabe2de97 , identity 43665->43665 state established ifindex lxc54076bf7fdd8 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp ACK
-> endpoint 2244 flow 0xabe2de97 , identity 43665->43665 state established ifindex lxc54076bf7fdd8 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp ACK
-> endpoint 2244 flow 0xabe2de97 , identity 43665->43665 state established ifindex lxc54076bf7fdd8 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp ACK
-> endpoint 2244 flow 0xabe2de97 , identity 43665->43665 state established ifindex lxc54076bf7fdd8 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp ACK
-> endpoint 2244 flow 0xabe2de97 , identity 43665->43665 state established ifindex lxc54076bf7fdd8 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp ACK, FIN
-> endpoint 2244 flow 0xabe2de97 , identity world->43665 state established ifindex 0 orig-ip 169.254.42.1: 169.254.42.1:51850 -> 10.244.1.210:8080 tcp ACK, FIN
```
